### ISSUE TITLE
Signatory does not seem to work with the P-256 algorithm 

### ISSUE THREAD

**Issue by amani1104** [ecadlabs/signatory#268]
I tested Signatory with the P-256 algorithm and imported a respective key into a YubiHSM 2:

`signatory-cli import --config ./signatory.yaml --vault yubi p2esXXX`

I then imported the key into the octez-client: 

`tezos import secret key test http://x.x.x.x:6732/tz3xxx`

Test if the key was imported in the YubiHSM:

```
signatory-cli list

INFO[0000] Initializing vault                            vault=yubihsm vault_name=yubi
Public Key Hash:    tz3xxx
Vault:              YubiHSM
ID:                 9013
Active:             true
Allowed Requests:   [generic]
Allowed Operations: [reveal transaction]
```

Output from YubiHSM Shell: 

```
Found 6 object(s)
id: 0x0001, type: authentication-key, sequence: 1
id: 0x112a, type: asymmetric-key, sequence: 0
id: 0x17e2, type: asymmetric-key, sequence: 0
--> id: 0x9013, type: asymmetric-key, sequence: 0
id: 0xd706, type: asymmetric-key, sequence: 0
id: 0xfb59, type: asymmetric-key, sequence: 0
```

I then funded the address with some Tez and tried to make a transaction from it. The error I get in Signatory is the following: 

```
Jan 05 15:25:23 signer signatory[10424]: time="2023-01-05T15:25:23Z" level=info msg="GET /authorized_keys" duration="39.354µs" hostname="x.x.x.x:6732" method=GET path=/authorized_keys start_time="2023-01-05T15:25:23Z" status=200
Jan 05 15:25:23 signer signatory[10424]: time="2023-01-05T15:25:23Z" level=info msg="Requesting signing operation" ops="map[reveal:1 transaction:1]" ops_total=2 pkh=tz3xxx request=generic vault=YubiHSM vault_name="127.0.0.1:12345/1"
Jan 05 15:25:23 signer signatory[10424]: time="2023-01-05T15:25:23Z" level=info msg="About to sign raw bytes" ops="map[reveal:1 transaction:1]" ops_total=2 pkh=tz3xxx raw=039ef5f10d0c3xxx>
Jan 05 15:25:23 signer signatory[10424]: 2023/01/05 15:25:23 http: panic serving x.x.x.x:49146: runtime error: invalid memory address or nil pointer dereference
Jan 05 15:25:23 signer signatory[10424]: goroutine 1094 [running]:
Jan 05 15:25:23 signer signatory[10424]: net/http.(*conn).serve.func1()
Jan 05 15:25:23 signer signatory[10424]:         /usr/local/go/src/net/http/server.go:1825 +0xbf
Jan 05 15:25:23 signer signatory[10424]: panic({0xd2de20, 0x17725a0})
Jan 05 15:25:23 signer signatory[10424]:         /usr/local/go/src/runtime/panic.go:844 +0x258
Jan 05 15:25:23 signer signatory[10424]: github.com/ecadlabs/signatory/pkg/cryptoutils.CanonizeECDSASignature(0xc0002e8700)
Jan 05 15:25:23 signer signatory[10424]:         /go/src/github.com/ecadlabs/signatory/pkg/cryptoutils/crypto.go:24 +0x2e2
Jan 05 15:25:23 signer signatory[10424]: github.com/ecadlabs/signatory/pkg/cryptoutils.CanonizeSignature(...)
Jan 05 15:25:23 signer signatory[10424]:         /go/src/github.com/ecadlabs/signatory/pkg/cryptoutils/crypto.go:44
Jan 05 15:25:23 signer signatory[10424]: github.com/ecadlabs/signatory/pkg/signatory.(*Signatory).Sign(0xc00051b320, {0x11edab8?, 0xc0008c2c30}, 0xc0005234c0)
Jan 05 15:25:23 signer signatory[10424]:         /go/src/github.com/ecadlabs/signatory/pkg/signatory/signatory.go:288 +0x1605
Jan 05 15:25:23 signer signatory[10424]: github.com/ecadlabs/signatory/pkg/server.(*Server).signHandler(0xc000278780, {0x7f9418346b98, 0xc00056e2b8}, 0xc0006a6b00)
Jan 05 15:25:23 signer signatory[10424]:         /go/src/github.com/ecadlabs/signatory/pkg/server/server.go:117 +0x5e4
Jan 05 15:25:23 signer signatory[10424]: net/http.HandlerFunc.ServeHTTP(0xc0008c2bd0?, {0x7f9418346b98?, 0xc00056e2b8?}, 0x203000?)
Jan 05 15:25:23 signer signatory[10424]:         /usr/local/go/src/net/http/server.go:2084 +0x2f
Jan 05 15:25:23 signer signatory[10424]: github.com/ecadlabs/signatory/pkg/server.(*Logging).Handler.func1({0x11ed0f8?, 0xc0006d2380}, 0xc0006a6b00)
Jan 05 15:25:23 signer signatory[10424]:         /go/src/github.com/ecadlabs/signatory/pkg/server/middleware.go:82 +0x190
Jan 05 15:25:23 signer signatory[10424]: net/http.HandlerFunc.ServeHTTP(0xc0006a6a00?, {0x11ed0f8?, 0xc0006d2380?}, 0xe200a0?)
Jan 05 15:25:23 signer signatory[10424]:         /usr/local/go/src/net/http/server.go:2084 +0x2f
Jan 05 15:25:23 signer signatory[10424]: github.com/gorilla/mux.(*Router).ServeHTTP(0xc0000e6000, {0x11ed0f8, 0xc0006d2380}, 0xc0006a6900)
Jan 05 15:25:23 signer signatory[10424]:         /go/pkg/mod/github.com/gorilla/mux@v1.7.3/mux.go:212 +0x202
Jan 05 15:25:23 signer signatory[10424]: net/http.serverHandler.ServeHTTP({0x11eb030?}, {0x11ed0f8, 0xc0006d2380}, 0xc0006a6900)
Jan 05 15:25:23 signer signatory[10424]:         /usr/local/go/src/net/http/server.go:2916 +0x43b
Jan 05 15:25:23 signer signatory[10424]: net/http.(*conn).serve(0xc000420140, {0x11edab8, 0xc0007c4840})
Jan 05 15:25:23 signer signatory[10424]:         /usr/local/go/src/net/http/server.go:1966 +0x5d7
Jan 05 15:25:23 signer signatory[10424]: created by net/http.(*Server).Serve
Jan 05 15:25:23 signer signatory[10424]:         /usr/local/go/src/net/http/server.go:3071 +0x4db
```

The setup is the following: 

1. YubiHSM 2, YubiHSM Connector and Signatory on a Intel NUC Core i3, Ubuntu 22.04, 16 GB RAM, 256 GB
2. Octez Node and Client on a AMD Ryzen 7 5800X 8x3.8 GHz, Ubuntu 20.04, 64 GB RAM, 1 TB

The setup works fine with a ED25519 key, but not with a P-256 key, although the docs state it should be supported on a YubiHSM2.

**Comment by AbineshECAD** [comment]
@amani1104 I'm looking into this issue. What is the signatory version you are using?

**Comment by amani1104** [comment]
@AbineshECAD Thanks for looking into it. I tested v0.4.0-beta and v1.0.0-beta1. Both have the same issue.

**Comment by AbineshECAD** [comment]
@amani1104 Thanks. Is the above crash output from v1.0.0-beta1? if not could you please share the crash output for v1.0.0-beta1?

**Comment by amani1104** [comment]
Just checked all the logs. There is no difference in the crash output for both versions.

**Comment by amani1104** [comment]
But here is the crash output from v1.0.0-beta1: 

```
Jan 05 20:36:29 signer signatory[14309]: time="2023-01-05T20:36:29Z" level=info msg="Requesting signing operation" ops="map[transaction:1]" ops_total=1 pkh=tz3xxx request=generic vault=YubiHSM vault_name="127.0.0.1:12345/1"
Jan 05 20:36:29 signer signatory[14309]: time="2023-01-05T20:36:29Z" level=info msg="About to sign raw bytes" ops="map[transaction:1]" ops_total=1 pkh=tz3xxx raw=03da6259xxx request=generic vault=YubiHSM vault_name="127.0.0.1:12345/1"
Jan 05 20:36:29 signer signatory[14309]: 2023/01/05 20:36:29 http: panic serving x.x.x.x:48890: runtime error: invalid memory address or nil pointer dereference
Jan 05 20:36:29 signer signatory[14309]: goroutine 131 [running]:
Jan 05 20:36:29 signer signatory[14309]: net/http.(*conn).serve.func1()
Jan 05 20:36:29 signer signatory[14309]:         /usr/local/go/src/net/http/server.go:1825 +0xbf
Jan 05 20:36:29 signer signatory[14309]: panic({0xd2f200, 0x17745a0})
Jan 05 20:36:29 signer signatory[14309]:         /usr/local/go/src/runtime/panic.go:844 +0x258
Jan 05 20:36:29 signer signatory[14309]: github.com/ecadlabs/signatory/pkg/cryptoutils.CanonizeECDSASignature(0xc00027a180)
Jan 05 20:36:29 signer signatory[14309]:         /go/src/github.com/ecadlabs/signatory/pkg/cryptoutils/crypto.go:24 +0x2e2
Jan 05 20:36:29 signer signatory[14309]: github.com/ecadlabs/signatory/pkg/cryptoutils.CanonizeSignature(...)
Jan 05 20:36:29 signer signatory[14309]:         /go/src/github.com/ecadlabs/signatory/pkg/cryptoutils/crypto.go:44
Jan 05 20:36:29 signer signatory[14309]: github.com/ecadlabs/signatory/pkg/signatory.(*Signatory).Sign(0xc0002c9810, {0x11ef8b8?, 0xc000793140}, 0xc000423d10)
Jan 05 20:36:29 signer signatory[14309]:         /go/src/github.com/ecadlabs/signatory/pkg/signatory/signatory.go:389 +0x16e5
Jan 05 20:36:29 signer signatory[14309]: github.com/ecadlabs/signatory/pkg/server.(*Server).signHandler(0xc0002d2dc0, {0x7f7d8d6d3558, 0xc000099128}, 0xc0005fd200)
Jan 05 20:36:29 signer signatory[14309]:         /go/src/github.com/ecadlabs/signatory/pkg/server/server.go:123 +0x624
Jan 05 20:36:29 signer signatory[14309]: net/http.HandlerFunc.ServeHTTP(0xc0007930e0?, {0x7f7d8d6d3558?, 0xc000099128?}, 0x203000?)
Jan 05 20:36:29 signer signatory[14309]:         /usr/local/go/src/net/http/server.go:2084 +0x2f
Jan 05 20:36:29 signer signatory[14309]: github.com/ecadlabs/signatory/pkg/server.(*Logging).Handler.func1({0x11eeef8?, 0xc00072e540}, 0xc0005fd200)
Jan 05 20:36:29 signer signatory[14309]:         /go/src/github.com/ecadlabs/signatory/pkg/server/middleware.go:82 +0x190
Jan 05 20:36:29 signer signatory[14309]: net/http.HandlerFunc.ServeHTTP(0xc0005fd100?, {0x11eeef8?, 0xc00072e540?}, 0xe219a0?)
Jan 05 20:36:29 signer signatory[14309]:         /usr/local/go/src/net/http/server.go:2084 +0x2f
Jan 05 20:36:29 signer signatory[14309]: github.com/gorilla/mux.(*Router).ServeHTTP(0xc00022a3c0, {0x11eeef8, 0xc00072e540}, 0xc0005fd000)
Jan 05 20:36:29 signer signatory[14309]:         /go/pkg/mod/github.com/gorilla/mux@v1.7.3/mux.go:212 +0x202
Jan 05 20:36:29 signer signatory[14309]: net/http.serverHandler.ServeHTTP({0x11ece30?}, {0x11eeef8, 0xc00072e540}, 0xc0005fd000)
Jan 05 20:36:29 signer signatory[14309]:         /usr/local/go/src/net/http/server.go:2916 +0x43b
Jan 05 20:36:29 signer signatory[14309]: net/http.(*conn).serve(0xc0002980a0, {0x11ef8b8, 0xc0004e08d0})
Jan 05 20:36:29 signer signatory[14309]:         /usr/local/go/src/net/http/server.go:1966 +0x5d7
Jan 05 20:36:29 signer signatory[14309]: created by net/http.(*Server).Serve
Jan 05 20:36:29 signer signatory[14309]:         /usr/local/go/src/net/http/server.go:3071 +0x4db
```

**Comment by amani1104** [comment]
Here is the error log from tezos-client, which may be helpful: 

```
octez-client transfer 0.1 from test to tz1xxx -D

Node is bootstrapped.
Estimated gas: 1000.040 units (will add 0 for safety)
Estimated storage: no bytes added
Rpc request failed:
   - meth: POST
   - uri: http://x.x.x.x:6732/keys/tz3xxx
   - error: Unable to connect to the node: "Server closed connection prematurely."
Fatal error:
  transfer simulation failed
```

**Comment by AbineshECAD** [comment]
> Just checked all the logs. There is no difference in the crash output for both versions.

@amani1104 The line numbers of the code are different which helps to be more accurate.

**Comment by AbineshECAD** [comment]
Update:
@amani1104 The crash ahppens because the library returns `nil` as it does not support this curve. We already faced this issue and updated the library.
### https://github.com/ecadlabs/signatory/issues/231

We are trying to test your scenario. Will update once it is done.

**Comment by AbineshECAD** [comment]
Update:
@amani1104 We tried your scenario with the latest code and still hit the issue you are facing. I don't have a YubiHSM with me, but I got a VM today with YubiHSM attached. I will update after some debugging today.

**Comment by AbineshECAD** [comment]
@amani1104 I Found how and where the nil pointer is coming from, and I'm working on a fix. I will update once the fix is tested and ready.

**Comment by AbineshECAD** [comment]
@amani1104 I tested my fix yesterday, and now the signing doesn't crash. We will merge it and let you know about the release update for a new image.

**Comment by amani1104** [comment]
@AbineshECAD That's great news! Thanks for the effort!

**Comment by amani1104** [comment]
I've seen that you included the fix in v1.0.0-beta2. I can confirm that it works now and that the issue is solved!

**Comment by jevonearth** [comment]
Closing, this fix has been shipped in [v1.0.0-beta2](https://github.com/ecadlabs/signatory/releases/tag/v1.0.0-beta2).

Thank you again for the report @amani1104
